File: algorithms/[1035]不相交的线/solution.ts
// topics = ['动态规划']

/**
 * Dynamic Programming
 * time O(mn), space O(mn), m 和 n 分别为 nums1 和 nums2 的长度
 */
function maxUncrossedLines(nums1: number[], nums2: number[]): number {
  const m = nums1.length;
  const n = nums2.length;
  // dp[i][j] 表示 nums1 前 i 个元素和 nums2 前 j 个元素能绘制的最多不相交的线数
  const dp = Array.from({ length: m + 1 }, () => new Array<number>(n + 1).fill(0));

  for (let i = 0; i <= m; i += 1) {
    dp[i][1] = nums1.slice(0, i).includes(nums2[0]) ? 1 : 0;
  }

  for (let j = 0; j <= n; j += 1) {
    dp[1][j] = nums2.slice(0, j).includes(nums1[0]) ? 1 : 0;
  }

  for (let i = 2; i <= m; i += 1) {
    for (let j = 2; j <= n; j += 1) {
      if (nums1[i - 1] == nums2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }

  return dp[m][n];
}

File: algorithms/[1035]不相交的线/solution.py
# topics = ['动态规划']

from typing import List


class Solution:
    def maxUncrossedLines(self, nums1: List[int], nums2: List[int]) -> int:
        """
        Dynamic Programming
        time O(mn), space O(mn), m 和 n 分别为 nums1 和 nums2 的长度
        """
        m, n = len(nums1), len(nums2)
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        for i, num1 in enumerate(nums1):
            for j, num2 in enumerate(nums2):
                if num1 == num2:
                    dp[i + 1][j + 1] = dp[i][j] + 1
                else:
                    dp[i + 1][j + 1] = max(dp[i][j + 1], dp[i + 1][j])

        return dp[m][n]
